drivers/custom_imports_boston_community_of_origin/app/models/custom_imports_boston_community_of_origin/import_file.rb

```diff
@@ -0,0 +1,173 @@
+        'Project Start Date' => 'do_not_import', # Use enrollment start date
+        'Project Exit Date' => 'do_not_import', # Use enrollment exit date
+        'Program ID' => 'do_not_import', # Use enrollment project id
+        'Name' => 'do_not_import', # Use project name
+        'City' => 'city',
+        'State' => 'state',
+        'Zip Code' => 'zip_code',
+        'How long has it been since you stayed in that community?' => 'length_of_time',
+        'Reporting Period Start Date' => 'do_not_import',
+        'Reporting Period End Date' => 'do_not_import',
+        'Client geolocation Location' => 'geolocation_location',
+      }
+    end
+
+    def post_process
+      update(status: 'processing')
+      # Records touched in this import
+      self.class.where(updated_at: started_at .. Time.current).
+        preload(enrollment: :project, client: :destination_client).
+        find_in_batches do |batch|
+        location_batch = []
+        batch.each do |row|
+          next unless row.client.present?
+
+          lat, lon = location(row)
+          location_batch << row.build_client_location(
+            client_id: row.client.destination_client.id,
+            located_on: contact_date(row),
+            lat: lat,
+            lon: lon,
+            collected_by: row&.enrollment&.project&.name,
+          )
+        end
+        ::ClientLocationHistory::Location.import!(
+          location_batch,
+          on_duplicate_key_update: {
+            conflict_target: [:source_id, :source_type],
+            columns: [:client_id, :located_on, :lat, :lon, :collected_by],
+          },
+        )
+      end
+
+      update(status: 'complete', completed_at: Time.current)
+      delayed_enrollment_location_histories
+    end
+
+    def delayed_enrollment_location_histories
+      ::GrdaWarehouse::Hud::Enrollment.delay.maintain_location_histories
+      Delayed::Worker.new.work_off if Rails.env.test?
+    end
+
+    def contact_date(row)
+      # We know that they were somewhere else before the enrollment, so use that date?
+      row.enrollment.EntryDate
+
+      # If we need better data, collected_on is the enrollment approximate date homeless (DateToStreetESSH)
+      # return nil unless row.collected_on.present?
+      #
+      # case row.length_of_time
+      # when 'Less than 1 week ago'
+      #   collected_on - 1.week.ago
+      # when 'About 1 month ago'
+      #   collected_on - 1.months.ago
+      # when 'Less than 6 months ago'
+      #   collected_on - 5.months.ago
+      # when 'More than 6 months ago'
+      #   collected_on - 11.months.ago
+      # when 'More than 1 year ago'
+      #   collected_on - 13.months.ago
+      # else
+      #
+      # end
+    end
+
+    # Geolocate based on available information.
+    # Search order: provided geolocation, zipcode, city, and then state
+    #
+    # @return [Array<Float>] [latitude, longitude] if location found, nil otherwise
+    def location(row)
```

Comment:
I feel like this should be able to short-circuit if the location has been calculated before (we can't call Nominatim too many times.)  To do that I think we need to store the lat/lon and address info used to calculate it elsewhere so we can look those up locally if we see them again.

Reply:
I added some caching to avoid duplicate lookups.

Reply:
Hmm, I might be missing something, but I meant long-term caching (like in the DB.)  Once we know the geolocation for a place, we shouldn't need to re-fetch it. (Most towns don't move.)

It looks like in `hmis_client.rb` and `grda_warehouse/hud/enrollment.rb` we're caching the result in redis for 6 weeks.  Not sure that's long enough, but it's better than per-load.  That said, maybe I'm missing something.  I also seem to have memories that must have been from a previous lifetime because I swear we had a table where we cached these based on query so we wouldn't ask Nominatim for the same query.

Reply:
Move geocoding to an async scheduled task.  
Run every 4 hours, check/set advisory lock so we only run one at a time.  
Cache query result in local `places` table 
Update GrdaWarehouse::Hud::Enrollment#address_lat_lon to use new caching mechanism.
`app/models/grda_warehouse/hmis_client.rb` also uses Nominatim for last_permanent_zip, let's bring it inline as well.